Tweak (facial) hair in preferences and icons

PsiOmegaDelta opened the pull request:
(Facial) Hair is now sorted lexically in the character setup.
Drastically reduces copy-paste.

Techhead0 commented on code/modules/client/preference_setup/general/03_body.dm:
```diff
@@ -705,3 +669,27 @@ var/global/list/valid_bloodtypes = list("A+", "A-", "B+", "B-", "AB+", "AB-", "O
+		var/datum/sprite_accessory/S = hair_styles_list[hairstyle]
+		if(!(mob_species.get_bodytype() in S.species_allowed))
+			continue
+		ADD_SORTED(L, hairstyle, /proc/cmp_text_asc)
```
Isn't sorting on every request a little computationally expensive to be doing repeatedly? A player might go through several hair styles before finding one they like.

PurpleMartinJCK replied:
You could make a list when species is initialised as valid hairstyles won't change.

PurpleMartinJCK commented on code/_helpers/icons.dm:
```diff
@@ -628,181 +628,203 @@ proc/ColorTone(rgb, tone)
+	// We start with a blank canvas, otherwise some icon procs crash silently
+	var/icon/flat = icon('icons/effects/effects.dmi', "nothing") // Final flattened icon
+	if(!A)
+		return flat
```
This should probably show an error of some kind or cause a runtime as it should be passed something.

PurpleMartinJCK commented on code/_helpers/icons.dm:
```diff
@@ -628,181 +628,203 @@ proc/ColorTone(rgb, tone)
+		return flat
+	var/noIcon = FALSE
+
+	if(start)
```
What does "start" mean? When should it be set to FALSE?

PsiOmegaDelta replied:
Who knows.

PurpleMartinJCK replied:
Could you find out?

PsiOmegaDelta replied:
Nope.

PurpleMartinJCK commented on code/_helpers/icons.dm:
```diff
@@ -628,181 +628,203 @@ proc/ColorTone(rgb, tone)
+// Creates a single icon from a given /atom or /image.  Only the first argument is required.
+/proc/getFlatIcon(image/A, defdir, deficon, defstate, defblend, start = TRUE, no_anim = FALSE)
+	// We start with a blank canvas, otherwise some icon procs crash silently
+	var/icon/flat = icon('icons/effects/effects.dmi', "nothing") // Final flattened icon
```
What icon size is nothing? Does it end up cutting off larger sprites?

PsiOmegaDelta replied:
32x32 as it always was. This line is practically identical to the original.
Don't know.

PurpleMartinJCK replied:
You should test it and let us know

PsiOmegaDelta replied:
Any particular reason why?

PurpleMartinJCK commented on code/_helpers/icons.dm:
```diff
@@ -628,181 +628,203 @@ proc/ColorTone(rgb, tone)
+
+	if(A.color)
+		flat.Blend(A.color, ICON_MULTIPLY)
+	if(A.alpha < 255)
```
Alpha composition isn't multiply afaik. Is this meant to be an approximation?

PsiOmegaDelta replied:
Perhaps.

PurpleMartinJCK commented on code/modules/client/preference_setup/general/03_body.dm:
```diff
@@ -366,23 +347,16 @@ var/global/list/valid_bloodtypes = list("A+", "A-", "B+", "B-", "AB+", "AB-", "O
 			return TOPIC_NOACTION
 		var/new_hair = input(user, "Choose your character's hair colour:", CHARACTER_PREFERENCE_INPUT_TITLE, rgb(pref.r_hair, pref.g_hair, pref.b_hair)) as color|null
-		if(new_hair && has_flag(mob_species, HAS_HAIR_COLOR) && CanUseTopic(user))
+		if(new_hair && has_flag(mob_species, HAS_HAIR_COLOR) && has_flag(mob_species, HAS_HAIR_COLOR) && CanUseTopic(user))
```
Checking the same flag twice?

PsiOmegaDelta committed:
Tweak (facial) hair in preferences

(Facial) Hair is now sorted lexically in the character setup.
Drastically reduces copy-paste.

Techhead0 reviewed (approved):
I guess get flat icon changes are moving to another PR.

PsiOmegaDelta commented:
I don't really care enough to bother with icons.
